python/iceberg/core/filesystem/s3_filesystem.py

```diff
@@ -0,0 +1,274 @@
+        S3_CLIENT[ROLE_ARN]["client"] = AUTOREFRESH_SESSION.client("s3")
+
+    return S3_CLIENT.get(ROLE_ARN, dict()).get(obj)
+
+
+def refresh_sts_session_keys():
+    params = {"RoleArn": ROLE_ARN,
+              "RoleSessionName": "iceberg_python_client_{}".format(int(time.time() * 1000.00))}
+
+    sts_creds = BOTO_STS_CLIENT.assume_role(**params).get("Credentials")
+    credentials = {"access_key": sts_creds.get("AccessKeyId"),
+                   "secret_key": sts_creds.get("SecretAccessKey"),
+                   "token": sts_creds.get("SessionToken"),
+                   "expiry_time": sts_creds.get("Expiration").isoformat()}
+    return credentials
+
+
+def url_to_bucket_key_name_tuple(url):
+    parsed_url = urlparse(url)
+    return parsed_url.netloc, parsed_url.path[1:], parsed_url.path.split("/")[-1]
+
+
+class S3FileSystem(FileSystem):
+    fs_inst = None
+
+    @staticmethod
+    def get_instance():
+        if S3FileSystem.fs_inst is None:
+            S3FileSystem()
+        return S3FileSystem.fs_inst
+
+    def __init__(self):
+        if S3FileSystem.fs_inst is None:
+            S3FileSystem.fs_inst = self
+
+    def set_conf(self, conf):
+        global ROLE_ARN
+        if conf is not None:
+            ROLE_ARN = conf.get("hive.aws_iam_role", "default")
+
+    def exists(self, path):
+        bucket, key, _ = url_to_bucket_key_name_tuple(S3FileSystem.normalize_s3_path(path))
+
+        res = get_s3(obj="client").list_objects_v2(Bucket=bucket,
+                                                   Prefix=key)
+        for obj in res.get('Contents', []):
+            return True
+
+        return False
+
+    def open(self, path, mode='rb'):
+        return S3File(path, mode=mode)
+
+    def delete(self, path):
+        bucket, key, _ = url_to_bucket_key_name_tuple(S3FileSystem.normalize_s3_path(path))
+        get_s3().Object(bucket_name=bucket,
+                        key=key).delete()
+
+    def stat(self, path):
+        is_dir = False
+        st = dict()
+        length = -1
+        try:
+            st = self.info(S3FileSystem.normalize_s3_path(path))
+        except RuntimeError:
+            # must be a directory or subsequent du will fail
+            du = self.du(S3FileSystem.normalize_s3_path(path))
+            if len(du) > 0:
+                is_dir = True
+
+            length = sum([val for key, val in du("s3://netflix-dataplatform-code/tgooch/test_write2.txt").items()])
+
+        return FileStatus(path=path, length=st.get("ContentLength", length), is_dir=is_dir,
+                          blocksize=None, modification_time=st.get("LastModified"), access_time=None,
+                          permission=None, owner=None, group=None)
+
+    @staticmethod
+    def du(url):
+        bucket, key, _ = url_to_bucket_key_name_tuple(url)
+        kwargs = {"Bucket": bucket,
+                  "Prefix": key}
+        du_items = {}
+        while True:
+            resp = get_s3().list_objects_v2(**kwargs)
```

Comment:
We probably want to be using the `Delimiter` argument for the v2 calls otherwise we are calculating the size for prefix, not a directory.

Reply:
I added the delimiter to the kwargs here, although I'm not sure that's quite right, the linux du command will list sub-directories and folders by default. With delimiter set, it will not here.  I'll share a notebook with you and take a look and let me know what you think